### PR TITLE
Skip submissions if any have been graded in next submitter view

### LINKED ISSUE #1196: The next submitter link for manual assessment
After merging the PR #1167 for the issue #1110, we found bugs in the new implementation. The pull request #1167 includes more comments about the details.

* The selected next submitter may be such a student who has already received manual grading on a submission other than the latest one. A student should be considered manually graded once any one of his/her submissions has been manually graded. The "next submitter" link should not repeat a student who has been manually graded.
* The "next submitter" link may repeat the same student who was already being inspected. This happens when the student's earlier submission (but not the latest) has been manually graded even though other students have not received any manual grading.
* When all students have received manual grading, the "next submitter" link must stop cycling between students like the old implementation had done. The old code has an if statement with the comment "There are no more unassessed submitters". It redirects back to the submission list.

_Originally posted by @Mankro in https://github.com/apluslms/a-plus/pull/1167#pullrequestreview-1474236719_

### PR DESCRIPTION
**Pull request by EerikSaksi**
Fixes #1196

# Description

**What?**

Submitters whose older work was already graded weren't skipped when pressing the "next submitter" link in manual assesment.

**Why?**
We shouldn't show a user whose work has already been graded even if they have later work.

**How?**

Rather than checking if the most recent submission has been graded, we skip submitters based on if any of their submissions for this exercise has been graded

Fixes #1196 

# Testing

**What type of test did you run?**

- [ ] Accessibility test using the [WAVE](https://wave.webaim.org/extension/) extension.
- [X] Django unit tests.
- [ ] Selenium tests.
- [ ] Other test. *(Add a description below)*
- [X] Manual testing.

In addition to unit tests I ensured that after grading the latest assesment the next submitter link takes you correctly. I also manually added a grader to the older work of the first user and ensured that this user was skipped even if their latest submission wasn't graded.

**Did you test the changes in**

- [ ] Chrome
- [X] Firefox
- [ ] This pull request cannot be tested in the browser.

**Think of what is affected by these changes and could become broken**

# Translation

- [ ] Did you modify or add new strings in the user interface? ([Read about how to create translation](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations))

# Programming style

- [X] Did you follow our [style guides](https://apluslms.github.io/contribute/styleguides/)?
- [ ] Did you use Python type hinting in all functions that you added or edited? ([type hints](https://docs.python.org/3/library/typing.html) for function parameters and return values)

# Have you updated the README or other relevant documentation?

- [ ] documents inside the doc directory.
- [ ] README.md.
- [ ] Aplus Manual.
- [ ] Other documentation (mention below which documentation).

# Is it Done?

- [ ] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

*Clean up your git commit history before submitting the pull request!*

### CODE REVIEW

File: exercise/staff_views.py [hunk 1]
```diff
@@ -329,6 +329,8 @@ def get(self, request: HttpRequest, *args: Any, **kwargs: Any) -> HttpResponse:
         # submissions for this exercise
         submissions = Submission.objects.filter(exercise=self.exercise)
 
+        submitters_with_grading = submissions.exclude(grader=None).values_list('submitters__id', flat=True)
+
         # filter only the latest submissions
         submissions = (submissions.annotate(latest=Max('submission_time'))
```
Comment (Mankro) [inline exercise/staff_views.py:335]: I don't think the `annotate`call for the `latest` submissions is working correctly. It doesn't seem to find the latest submission for each submitter separately.

**Review by Mankro — CHANGES_REQUESTED** [review]
I think the query for the latest submissions of each submitter is still not working correctly.

Remember also that a submission may have multiple submitters and those cases should also be tested. The current unit tests use too few submissions and different submitters to reveal significant bugs.

I am pushing a new commit that fixes minor issues in the current unit tests.

**Comment by Mankro** [comment]
I looked at the SQL query produced by the "submissions annotate latest` statement.

```python
(submissions.annotate(latest=Max('submission_time'))
    .filter(submission_time=F('latest'))
    .order_by('latest')))
```

This GROUP BY part looks suspicious since it is not grouping by submitters. See the whole query below.
```sql
GROUP BY "exercise_submission"."id"
HAVING "exercise_submission"."submission_time" = MAX("exercise_submission"."submission_time")
```

The whole SQL query:
```sql
SELECT "exercise_submission"."id",
       "exercise_submission"."submission_time",
       "exercise_submission"."hash",
       "exercise_submission"."exercise_id",
       "exercise_submission"."grader_id",
       "exercise_submission"."feedback",
       "exercise_submission"."assistant_feedback",
       "exercise_submission"."status",
       "exercise_submission"."grade",
       "exercise_submission"."grading_time",
       "exercise_submission"."late_penalty_applied",
       "exercise_submission"."force_exercise_points",
       "exercise_submission"."service_points",
       "exercise_submission"."service_max_points",
       "exercise_submission"."submission_data",
       "exercise_submission"."grading_data",
       "exercise_submission"."meta_data",
       MAX("exercise_submission"."submission_time") AS "latest"
  FROM "exercise_submission"
 INNER JOIN "exercise_submission_submitters"
    ON ("exercise_submission"."id" = "exercise_submission_submitters"."submission_id")
 WHERE ("exercise_submission"."exercise_id" = 61 AND "exercise_submission_submitters"."userprofile_id" = 7)
 GROUP BY "exercise_submission"."id"
HAVING "exercise_submission"."submission_time" = MAX("exercise_submission"."submission_time")
 ORDER BY "latest" ASC
```

I think the part for querying a single user (`"exercise_submission_submitters"."userprofile_id" = 7`) came from the previous submitter (`prev` GET query parameter):

```python
previous_submission_time = submissions.filter(submitters__id=previous_user_id).first().submission_time
```

**Comment by EerikSaksi** [comment]
I decided to try to try reimplement it by making small changes to the original query and managed to create a query which seems to provide the correct behavior for multiple users, and seems to group users based on their profile and sort them based on their earliest submission

```
submitters = (UserProfile.objects
		.filter(submissions__exercise=self.exercise)
		.annotate(
				count_assessed=Count(
						'submissions__id',
						filter=(Q(submissions__grader__isnull=False)),
				),
				latest_submission=Min('submissions__submission_time'),
		)
		.order_by('latest_submission'))
```

```
SELECT
  "userprofile_userprofile"."id",
  "userprofile_userprofile"."user_id",
  "userprofile_userprofile"."language",
  "userprofile_userprofile"."student_id",
  "userprofile_userprofile"."organization",
  COUNT("exercise_submission_submitters"."submission_id") FILTER (
    WHERE
      "exercise_submission"."grader_id" IS NOT NULL
  ) AS "count_assessed",
  MIN("exercise_submission"."submission_time") AS "latest_submission",
  "auth_user"."id",
  "auth_user"."password",
  "auth_user"."last_login",
  "auth_user"."is_superuser",
  "auth_user"."username",
  "auth_user"."first_name",
  "auth_user"."last_name",
  "auth_user"."email",
  "auth_user"."is_staff",
  "auth_user"."is_active",
  "auth_user"."date_joined"
FROM
  "userprofile_userprofile"
  LEFT OUTER JOIN "exercise_submission_submitters" ON (
    "userprofile_userprofile"."id" = "exercise_submission_submitters"."userprofile_id"
  )
  LEFT OUTER JOIN "exercise_submission" ON (
    "exercise_submission_submitters"."submission_id" = "exercise_submission"."id"
  )
  INNER JOIN "auth_user" ON (
    "userprofile_userprofile"."user_id" = "auth_user"."id"
  )
WHERE
  "exercise_submission"."exercise_id" = 25
GROUP BY
  "userprofile_userprofile"."id",
  "auth_user"."id"
ORDER BY
  "latest_submission" ASC
```

**Pull request merged**